poetry/puzzle/solver.py

```diff
@@ -84,6 +84,13 @@ def solve(self, use_latest=None):  # type: (...) -> List[Operation]
                     elif package.version != pkg.version:
                         # Checking version
                         operations.append(Update(pkg, package))
+                    elif (
+                        package.source_type == "directory"
+                        and pkg.source_type != "directory"
```

Comment:
I wonder if we can generalize this for any type of change in the dependency source type (`file` -> `directory` or `directory` -> `git` for instance)

Reply:
That seems like it ought to be the right answer, but it looks as though the detection of source type for installed packages is not currently reliable:

- so far as I can see, it is [only ever set to "git"](https://github.com/sdispater/poetry/blob/64e8af3b4f19d3991a983745cfd313d37c92e802/poetry/repositories/installed_repository.py#L38) (or nothing at all)
- and actually, even packages that are installed as "directory" are reported as "git", if it happens that the code is in a git repository (which is fairly likely to be the case).
  - I suppose by the time you get to doing `pip freeze`, the two cases are indistinguishable

Presumably in an ideal world the source_type of installed packages would be made reliable - but it's not clear to me how to achieve that.